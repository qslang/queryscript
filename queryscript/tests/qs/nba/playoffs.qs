import playoffs, teams from schema;

let normalized_games =
    SELECT
        TEAM_ID AS team,
        SEASON as season,
        strptime(GAME_DATE, '%Y-%m-%d') AS game_date,
        (SELECT TEAM_ID FROM playoffs i WHERE i.GAME_ID = o.GAME_ID and i.TEAM_ID != o.TEAM_ID) opponent,
        CASE WHEN WL='W' THEN 1 ELSE -1 END AS won
    from playoffs o;

-- TODO: Support GROUP BY ALL
-- https://github.com/sqlparser-rs/sqlparser-rs/issues/791
let series = SELECT
        season,
        team,
        opponent,
        SUM(won) > 0 AS winner,
        MIN(game_date) as first_game
    FROM normalized_games
    GROUP BY 1, 2, 3;
;

let series_ranked =
    select *, ROW_NUMBER() OVER (PARTITION BY (season, team) ORDER BY first_game) AS "round_number" FROM series
;

let first_round_bye =
    SELECT
        series_ranked.season, series_ranked.team, series_ranked.round_number
    FROM series_ranked
    JOIN series_ranked o ON series_ranked.opponent = o.team AND o.opponent = series_ranked.team AND series_ranked.season = o.season
    WHERE
        series_ranked.round_number = 1 AND o.round_number = 2
;

-- TODO: Use REPLACE syntax from DuckDB
-- https://github.com/sqlparser-rs/sqlparser-rs/issues/792
-- TODO: USING
let series_ranked_with_bye =
    SELECT
        series_ranked.* EXCLUDE (round_number),
        series_ranked.round_number + (CASE WHEN first_round_bye.team IS NULL THEN 0 ELSE 1 END) AS round_number
    FROM series_ranked LEFT JOIN first_round_bye ON series_ranked.season = first_round_bye.season AND series_ranked.team = first_round_bye.team
;

-- Let the 0th round be the "entry" to the playoffs, and the rest (e.g. 1st) be the winner.
let all_series =
    SELECT * EXCLUDE (round_number), 0 as round_number FROM series_ranked_with_bye WHERE round_number = 1
    UNION ALL
    SELECT * FROM series_ranked_with_bye WHERE winner=true
;

-- TODO: Placing the expression in parantheses makes it a "scalar subselect". I don't think we can
-- infer this automatically because it relies on the assumption that it should be treated as a scalar.
-- Perhaps we could infer this by inspecting the return type, or picking a more explicit syntax.
-- TODO: Infer the type of team
fn full_name(team bigint) {
    (SELECT full_name FROM teams WHERE team = teams.ID LIMIT 1)
}

-- TODO: Infer the type of n
fn n_to_last_round(n bigint) {
    -- TODO: The query we'd want to run doesn't parse, because `n` is checked to be a constant at parsing time.
    -- SELECT DISTINCT round_number FROM all_series ORDER BY round_number DESC LIMIT 1 OFFSET n
    (SELECT round_number FROM (
        SELECT round_number, ROW_NUMBER() OVER (ORDER BY round_number DESC) AS "rank" FROM all_series GROUP BY 1
    ) WHERE rank = n + 1)
}

-- Teams that showed up in the playoffs the most
SELECT full_name(team) full_name, COUNT(*) total FROM all_series GROUP BY 1 ORDER BY 2 DESC, 1 LIMIT 10;

-- Teams that showed up in the finals the most
SELECT full_name(team) full_name, COUNT(*) total FROM all_series
    WHERE round_number = n_to_last_round(1)
GROUP BY 1 ORDER BY 2 DESC LIMIT 10;

let winners =
    SELECT team, COUNT(*) total FROM all_series
        WHERE round_number = n_to_last_round(0)
    GROUP BY 1
;

-- TODO: Infer the type of team
fn is_winner(team bigint) {
    team IN (SELECT DISTINCT team FROM winners)
}

-- Teams that won the most
SELECT full_name(team), total FROM winners ORDER BY 2 DESC LIMIT 10;

-- Teams that showed up in the playoffs but never won
SELECT full_name(team), COUNT(*) total FROM all_series
    WHERE NOT is_winner(team)
GROUP BY 1 ORDER BY 2 DESC LIMIT 10;

-- Teams that showed up in the finals but never won
SELECT full_name(team), COUNT(*) total FROM all_series
    WHERE round_number = n_to_last_round(1)
    AND NOT is_winner(team)
GROUP BY 1 ORDER BY 2 DESC LIMIT 10;

-- Overall funnel
SELECT round_number, COUNT(DISTINCT team) teams FROM all_series GROUP BY 1 ORDER BY 1
WITH VIZ {
};


SELECT full_name(team) full_name, total FROM winners
WITH VIZ {
    mark: 'arc',
};
