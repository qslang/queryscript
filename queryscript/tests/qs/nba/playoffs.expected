{
    "compile_errors": [],
    "decls": {
        "let all_series": [{
        	season Int64,
        	team Int64,
        	opponent Int64,
        	winner Boolean,
        	first_game Timestamp(
            Microsecond,
            None,
        ),
        	round_number Int64,
        }],
        "let first_round_bye": [{
        	season Int64,
        	team Int64,
        	round_number Int64,
        }],
        "let full_name": λ {
        	team Int64,
        } -> Utf8,
        "let is_winner": λ {
        	team Int64,
        } -> Boolean,
        "let n_to_last_round": λ {
        	n Int64,
        } -> Int64,
        "let normalized_games": [{
        	team Int64,
        	season Int64,
        	game_date Timestamp(
            Microsecond,
            None,
        ),
        	opponent Int64,
        	won Int64,
        }],
        "let playoffs": External<[{
        	SEASON_ID Utf8,
        	TEAM_ID Int64,
        	TEAM_ABBREVIATION Utf8,
        	TEAM_NAME Utf8,
        	GAME_ID Utf8,
        	GAME_DATE Utf8,
        	MATCHUP Utf8,
        	WL Utf8,
        	MIN Int64,
        	FGM Int64,
        	FGA Float64,
        	FG_PCT Float64,
        	FG3M Int64,
        	FG3A Float64,
        	FG3_PCT Float64,
        	FTM Int64,
        	FTA Int64,
        	FT_PCT Float64,
        	OREB Float64,
        	DREB Float64,
        	REB Float64,
        	AST Float64,
        	STL Float64,
        	BLK Float64,
        	TOV Float64,
        	PF Float64,
        	PTS Int64,
        	PLUS_MINUS Int64,
        	VIDEO_AVAILABLE Int64,
        	SEASON Int64,
        	__index_level_0__ Int64,
        }]>,
        "let series": [{
        	season Int64,
        	team Int64,
        	opponent Int64,
        	winner Boolean,
        	first_game Timestamp(
            Microsecond,
            None,
        ),
        }],
        "let series_ranked": [{
        	season Int64,
        	team Int64,
        	opponent Int64,
        	winner Boolean,
        	first_game Timestamp(
            Microsecond,
            None,
        ),
        	round_number Int64,
        }],
        "let series_ranked_with_bye": [{
        	season Int64,
        	team Int64,
        	opponent Int64,
        	winner Boolean,
        	first_game Timestamp(
            Microsecond,
            None,
        ),
        	round_number Int64,
        }],
        "let teams": External<[{
        	id Int64,
        	full_name Utf8,
        	abbreviation Utf8,
        	nickname Utf8,
        	city Utf8,
        	state Utf8,
        	year_founded Int64,
        }]>,
        "let winners": [{
        	team Int64,
        	total Int64,
        }],
    },
    "queries": [
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "full_name",
                                type_: Atom(
                                    Utf8,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "total",
                                type_: Atom(
                                    Int64,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| full_name              | total |\n|------------------------|-------|\n| Los Angeles Lakers     | 98    |\n| Boston Celtics         | 77    |\n| San Antonio Spurs      | 76    |\n| Chicago Bulls          | 64    |\n| Houston Rockets        | 56    |\n| Utah Jazz              | 56    |\n| Detroit Pistons        | 55    |\n| Miami Heat             | 53    |\n| Portland Trail Blazers | 51    |\n| Oklahoma City Thunder  | 50    |",
            },
        ),
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "full_name",
                                type_: Atom(
                                    Utf8,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "total",
                                type_: Atom(
                                    Int64,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| full_name             | total |\n|-----------------------|-------|\n| Los Angeles Lakers    | 16    |\n| Boston Celtics        | 8     |\n| San Antonio Spurs     | 6     |\n| Miami Heat            | 6     |\n| Chicago Bulls         | 6     |\n| Golden State Warriors | 6     |\n| Detroit Pistons       | 5     |\n| Cleveland Cavaliers   | 5     |\n| Houston Rockets       | 4     |\n| Philadelphia 76ers    | 3     |",
            },
        ),
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "full_name(team)",
                                type_: Atom(
                                    Utf8,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "total",
                                type_: Atom(
                                    Int64,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| full_name(team)       | total |\n|-----------------------|-------|\n| Los Angeles Lakers    | 10    |\n| Chicago Bulls         | 6     |\n| San Antonio Spurs     | 5     |\n| Boston Celtics        | 4     |\n| Golden State Warriors | 4     |\n| Miami Heat            | 3     |\n| Detroit Pistons       | 3     |\n| Houston Rockets       | 2     |\n| Cleveland Cavaliers   | 1     |\n| Dallas Mavericks      | 1     |",
            },
        ),
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "full_name(team)",
                                type_: Atom(
                                    Utf8,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "total",
                                type_: Atom(
                                    Int64,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| full_name(team)        | total |\n|------------------------|-------|\n| Utah Jazz              | 56    |\n| Portland Trail Blazers | 51    |\n| Oklahoma City Thunder  | 50    |\n| Phoenix Suns           | 50    |\n| Indiana Pacers         | 46    |\n| Atlanta Hawks          | 42    |\n| New York Knicks        | 42    |\n| Denver Nuggets         | 37    |\n| Brooklyn Nets          | 34    |\n| Orlando Magic          | 27    |",
            },
        ),
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "full_name(team)",
                                type_: Atom(
                                    Utf8,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "total",
                                type_: Atom(
                                    Int64,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| full_name(team)        | total |\n|------------------------|-------|\n| Portland Trail Blazers | 2     |\n| Phoenix Suns           | 2     |\n| Brooklyn Nets          | 2     |\n| New York Knicks        | 2     |\n| Utah Jazz              | 2     |\n| Orlando Magic          | 2     |\n| Oklahoma City Thunder  | 2     |\n| Indiana Pacers         | 1     |",
            },
        ),
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "round_number",
                                type_: Atom(
                                    Int64,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "teams",
                                type_: Atom(
                                    Int64,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| round_number | teams |\n|--------------|-------|\n| 0            | 30    |\n| 1            | 30    |\n| 2            | 27    |\n| 3            | 21    |\n| 4            | 13    |",
            },
        ),
        Ok(
            TypedValue {
                type_: List(
                    Record(
                        [
                            Field {
                                name: "full_name",
                                type_: Atom(
                                    Utf8,
                                ),
                                nullable: true,
                            },
                            Field {
                                name: "total",
                                type_: Atom(
                                    Int64,
                                ),
                                nullable: true,
                            },
                        ],
                    ),
                ),
                value: "| full_name             | total |\n|-----------------------|-------|\n| Boston Celtics        | 4     |\n| Cleveland Cavaliers   | 1     |\n| Chicago Bulls         | 6     |\n| Dallas Mavericks      | 1     |\n| Golden State Warriors | 4     |\n| Houston Rockets       | 2     |\n| Los Angeles Lakers    | 10    |\n| Miami Heat            | 3     |\n| Milwaukee Bucks       | 1     |\n| Philadelphia 76ers    | 1     |\n| San Antonio Spurs     | 5     |\n| Toronto Raptors       | 1     |\n| Detroit Pistons       | 3     |",
            },
        ),
    ],
}